src/omnisharp/OmnisharpDownloader.ts

```diff
@@ -4,69 +4,96 @@
+
+        this.telemetryProps = {};
+        this.packageManager = new PackageManager(this.platformInfo, this.packageJSON);
     }
 
     public async DownloadAndInstallOmnisharp(version: string, serverUrl: string, installPath: string) {
         if (!version) {
             throw new Error('Invalid version');
         }
 
-        this.logger.append('Installing Omnisharp Packages...');
+        this.logger.appendLine('Installing Omnisharp Packages...');
         this.logger.appendLine();
         this.channel.show();
 
-        let statusObject = SetStatus();
-        let status = statusObject.Status;
-        let statusItem = statusObject.StatusItem;
-
-        let telemetryProps: any = {};
         let installationStage = '';
-        let platformInfo: PlatformInformation;
 
         if (this.reporter) {
             this.reporter.sendTelemetryEvent("AcquisitionStart");
         }
 
         try {
-            installationStage = 'getPlatformInfo';
-            platformInfo = await GetAndLogPlatformInformation(this.logger);
+            LogPlatformInformation(this.logger, this.platformInfo);
 
             installationStage = 'getPackageInfo';
             let packages: Package[] = GetPackagesFromVersion(version, this.packageJSON.runtimeDependencies, serverUrl, installPath);
 
             installationStage = 'downloadPackages';
-            let packageManager = new PackageManager(platformInfo, this.packageJSON);
+
             // Specify the packages that the package manager needs to download
-            packageManager.SetVersionPackagesForDownload(packages);
-            await GetDependenciesAndDownloadPackages(packages,status, platformInfo, packageManager, this.logger);
+            this.packageManager.SetVersionPackagesForDownload(packages);
+            await this.packageManager.DownloadPackages(this.logger, this.status, this.proxy, this.strictSSL);
 
             this.logger.appendLine();
 
             installationStage = 'installPackages';
-            await packageManager.InstallPackages(this.logger, status);
+            await this.packageManager.InstallPackages(this.logger, this.status);
 
             installationStage = 'completeSuccess';
         }
-
         catch (error) {
-            ReportInstallationError(this.logger, error, telemetryProps, installationStage);
+            ReportInstallationError(this.logger, error, this.telemetryProps, installationStage);
             throw error;// throw the error up to the server
         }
         finally {
-            SendInstallationTelemetry(this.logger, this.reporter, telemetryProps, installationStage, platformInfo, statusItem);
+            SendInstallationTelemetry(this.logger, this.reporter, this.telemetryProps, installationStage, this.platformInfo, this.statusItem);
+        }
+    }
+
+    public async GetLatestVersion(serverUrl: string, versionFilePathInServer): Promise<string> {
+        let installationStage = 'getLatestVersionInfoFile';
```

Comment:
In our current pattern, anyone who wants to add an installation stage and print errors on failure has to set up an identical try/catch/finally block. I wonder if we could add a `RunInstllationStage` helper that takes an async func that does the work, the stage name, and the logger and does all that for you.

Reply:
Do you see this as a regression in the source? There is a lot of refactoring to be done in this area of the code and I'd like to limit the changes in this PR lest it grows too big...

Reply:
Technically it's not a pattern this PR introduced, but this PR does create more instances of it. @akshita31 I'd be OK with getting this refactoring in a followup PR.